history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
```

Comment:
Actually, another way to think about the suggested changes is to have the history be ahead of the current epoch. Then it would be safe to clone the tree after the inserts. In other words, after ``STR_t`` is created and ``tree_t``, we create the new snapshot for ``t+1`` by cloning ``tree_t`` into ``tree_t+1``. Then it is safe to insert new leaves into ``tree_t+1`` as long as lookups are done in ``tree_t`` until ``STR_t+1`` is created.

Thus, ``UpdateHistory()`` could look something like this:
``
func (s *Snapshot) UpdateHistory() (nextSnapshot *Snapshot){
  s.tree.RecomputeHash()
  nextEp := s.currentSTR.epoch + s.epochInterval
  s.currentSTR = s.currentSTR.generateNextSTR(s.tree, nextEp)
 nextSnapshot = NewSnapshot(s.tree.clone())
 nextSnapshot.prev = s
 return
}
``
I know this isn't exactly true to the implementation of ``History`` and STRs but I hope you get the idea. What do you think?